integration-tests/tests/src/tests/counter.ts

```diff
@@ -0,0 +1,994 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { expect } from "chai";
+import Realm, { BSON, Counter, ObjectSchema, UpdateMode } from "realm";
+
+import { openRealmBeforeEach } from "../hooks";
+import { expectRealmDictionary, expectRealmList, expectRealmSet } from "../utils/expects";
+
+interface IWithCounter {
+  _id: BSON.ObjectId;
+  counter: Counter;
+}
+
+const WithCounterSchema: ObjectSchema = {
+  name: "WithCounter",
+  primaryKey: "_id",
+  properties: {
+    _id: { type: "objectId", default: () => new BSON.ObjectId() },
+    counter: "counter",
+  },
+};
+
+interface IWithNullableCounter {
+  _id: BSON.ObjectId;
+  nullableCounter?: Counter | null;
+}
+
+const WithNullableCounterSchema: ObjectSchema = {
+  name: "WithNullableCounter",
+  primaryKey: "_id",
+  properties: {
+    _id: { type: "objectId", default: () => new BSON.ObjectId() },
+    nullableCounter: "counter?",
+  },
+};
+
+interface IWithDefaultCounter {
+  counterWithDefault: Counter;
+}
+
+const WithDefaultCounterSchema: ObjectSchema = {
+  name: "WithDefaultCounter",
+  properties: {
+    counterWithDefault: { type: "int", presentation: "counter", default: 0 },
+  },
+};
+
+interface IWithRegularInt {
+  int: number;
+}
+
+const WithRegularIntSchema: ObjectSchema = {
+  name: "WithRegularInt",
+  properties: {
+    int: "int",
+  },
+};
+
+interface IWithMixed {
+  mixed: Realm.Types.Mixed;
+  list: Realm.List<Realm.Types.Mixed>;
+  dictionary: Realm.Dictionary<Realm.Types.Mixed>;
+  set: Realm.Set<Realm.Types.Mixed>;
+}
+
+const WithMixedSchema: ObjectSchema = {
+  name: "WithMixed",
+  properties: {
+    mixed: "mixed",
+    list: "mixed[]",
+    dictionary: "mixed{}",
+    set: "mixed<>",
+  },
+};
+
+function expectCounter(value: unknown): asserts value is Counter {
+  expect(value).to.be.instanceOf(Counter);
+}
+
+function expectKeys(dictionary: Realm.Dictionary, keys: string[]) {
+  expect(Object.keys(dictionary)).members(keys);
+}
+
+describe("Counter", () => {
+  openRealmBeforeEach({
+    schema: [
+      WithCounterSchema,
+      WithNullableCounterSchema,
+      WithDefaultCounterSchema,
+      WithRegularIntSchema,
+      WithMixedSchema,
+    ],
+  });
+
+  const initialValuesList = [-100, 0, 1.0, 1000] as const;
+
+  describe("Create and access", () => {
+    describe("Via 'realm.create()'", () => {
+      it("can create and access (input: number)", function (this: RealmContext) {
+        for (let i = 0; i < initialValuesList.length; i++) {
+          const input = initialValuesList[i];
+          const { counter } = this.realm.write(() => {
+            return this.realm.create<IWithCounter>(WithCounterSchema.name, {
+              counter: input,
+            });
+          });
+
+          const expectedNumObjects = i + 1;
```

Comment:
```suggestion
          const expectedNumObjects = initialValuesList.length;
```

Reply:
I know that the variable `i` is declared in the scope outside the loop (and each iteration has its own scope), but I always find it a bit hard to read.